### DISCUSSION TITLE
First steps using PIO: viable strategies?

### DISCUSSION THREAD

**Discussion by mdaeron** [micropython/micropython#10887] [RP2040 / Pico] [answered]
I'm trying to write a driver for ZACwire devices such as the [TSic 506F](https://docs.rs-online.com/e2fb/0900766b81690bb9.pdf) temperature sensor. One useful property of these sensors is that their nominal accuracy is excellent (±0.1 °C), which is necessary for some scientific applications (e.g., ensuring precise *and* accurate control of a chemical reaction's temperature). I've managed to write a usable driver for the [pyboard](https://github.com/mdaeron/micropython-zacwire) based on IRQs and a Timer, but I'd like to learn how to do this using a Pico W's PIO. I've looked at the documentation and some examples from [this thread](https://forum.micropython.org/viewtopic.php?t=11463#p62683), and I'm just starting to grasp the basics of PIO programming.

Before I go further, I have to decide on an overall strategy to divide work between the PIO and the main system. The ZACwire protocols encodes bits as a series of 20 low pulse durations ranging from 20 to 100 microseconds. From other's examples I see how I can measure each of these durations and push them to the ISR using `rp2.in_()`.

* Option A would be to add an IRQ that would tell the main system when to read this value and leave all book-keeping (storing the values in the correct sequence, and processing the result after 20 values have been read) to the main system. I'm not sure this really better than just using IRQs with calls to `ticks_us()`, as I did on the pyboard.
* Option B would be to have the PIO program somehow fill a buffer that could be read from the main system, and only create an IRQ when 20 pulses have been read. But in that case I still need to add some logic to detect when a bit sequence starts (each sequence is separated by ~97 milliseconds of high state, so I guess I could decrement one of the scratch buffers to detect when a high state lasts more than 300 microseconds).
* Option C would be to forget all about the PIO because it's not well suited to this use case.

I'm somewhat optimistic that option B is not unfeasible, and I sure hope it's not option C. Any general suggestions, or pointers to specific methods/functions that are well suited to this task, would be much appreciated. Of course, I'll add this to [micropython-zacwire](https://github.com/mdaeron/micropython-zacwire) if/when it works well.

**Comment by robert-hh** [comment]
You could have a look at this class: https://github.com/robert-hh/RP2040-Examples/tree/master/pulses
It simply can be used to time pulses coming in on a PIO pin. Similar other examples may be available at https://awesome-micropython.com/

**Reply by mdaeron** [reply]
Thanks @robert-hh, I just starred your repo perhaps 2 minutes ago! I'm reading the code and trying to wrap my head around it. It seems that with a TSic, `pulses.get()` blocks (as expected) for tens of milliseconds waiting for the next pulse train to start. I'm going to look at your DMA functions to see how I can fill a buffer as a background task, so that my main system only has to (1) check whether a buffer write is in progress, (2) read the buffer and (3) perform the conversion to temperature reradings (including parity check).

Thanks again!

**Comment by mdaeron** [comment, marked as answer]
Thanks to the examples kindly provided by @robert-hh, I got something working well on a Pico W. The driver is [here](https://github.com/mdaeron/micropython-zacwire/tree/master/rp2), and uses two state machines:

* The first one (`sm0`) reports the low pulses durations out of the TSic.
* The second state machine (`sm1`) raises an IRQ when it detects a high pulse longer than ~5 ms, which only happens in between temperature readings.
* The callback to that IRQ starts a DMA transfer which will copy the next twenty pulse durations reported by `sm0` and write them to a buffer array `dmabuf`. It also schedules the decoding of `dmabuf` into a temperature reading

Any comments or suggestions for improvements are welcome, and thanks again to @robert-hh.

**Update:** Due to #8431, the DMA approach is brittle in many situations. I've moved back to using two IRQs, which still works well. I'd going to DMA if/when this issue is resolved.
